File: twitter/twitter_api.go
package twitter

import (
	"io/ioutil"
	//"log"
	"encoding/json"
	"fmt"
	"internet-thoughts/sentiment"
	"net/http"
	"sort"
)

type TwitterMain struct {
	Meta Meta   `json:"meta"`
	Data []Data `json:"data"`
}

type Meta struct {
	Next_token string `json:"next_token"`
}
type Data struct {
	Id             string         `json:"id"`
	Text           string         `json:"text"`
	Public_Metrics Public_Metrics `json:"public_metrics"`
}

type Public_Metrics struct {
	Retweet_Count int `json:"retweet_count"`
	Reply_Count   int `json:"reply_count"`
	Like_Count    int `json:"like_count"`
	Quote_Count   int `json:"quote_count"`
}

func TwitterCall(searchTerm string, location string, numberOfTweets int) []TwitterMain {

	var retList []TwitterMain
	client := &http.Client{}
	nextToken := ""
	for numberOfTweets > 0 {
		curNum := 100
		if numberOfTweets < 100 {
			curNum = numberOfTweets
		}

		url := fmt.Sprintf("https://api.twitter.com/2/tweets/search/recent?query=%s&tweet.fields=geo,public_metrics&place.fields=country,name&max_results=%d", searchTerm, curNum)
		if nextToken != "" {
			url += "&next_token=" + nextToken
		}
		req, _ := http.NewRequest("GET", url, nil)
		req.Header.Set("Authorization", "Bearer AAAAAAAAAAAAAAAAAAAAAJDDNwEAAAAAvR58LgY1daTewd2C4htoRLLBxO4%3D2gqj5PSlNfMOnrSVwX8La6FNsBj6Rc5vJRbH81tWqkoLS4Lcej")

		resp, _ := client.Do(req)
		responseData, _ := ioutil.ReadAll(resp.Body)
		var twitterResponse TwitterMain
		json.Unmarshal(responseData, &twitterResponse)
		nextToken = twitterResponse.Meta.Next_token
		numberOfTweets -= curNum
		retList = append(retList, twitterResponse)

	}
	return retList
}

func Twitter_Most(data_arr []Data, n int) []sentiment.Buzz {
	var x []sentiment.Buzz
	sort.Slice(data_arr, func(i, j int) bool {
		return data_arr[i].Public_Metrics.Retweet_Count > data_arr[j].Public_Metrics.Retweet_Count
	})
	for i := 0; i < n; i++ {
		var tmp sentiment.Buzz
		tmp.Id = data_arr[i].Id
		x = append(x, tmp)

	}
	return x
}

File: python/twitter_api.py
import requests
import os
import json

bearerToken = ""




#by date

#by location

#by keyword

#number of tweets

#returns number of likes

#returns number of retweets

#
def create_url(searchTerm, location, start_time, end_time, numberOfTweets, next_token):
    query = searchTerm
    # Tweet fields are adjustable.
    # Options include:
    # attachments, author_id, context_annotations,
    # conversation_id, created_at, entities, geo, id,
    # in_reply_to_user_id, lang, non_public_metrics, organic_metrics,
    # possibly_sensitive, promoted_metrics, public_metrics, referenced_tweets,
    country_fields = ""

    start_field = ""
    end_field = ""
    # source, text, and withheld
    tweet_fields = "tweet.fields=geo,created_at,public_metrics&place.fields=country,name"

    if end_time:
    	end_field = "&end_time=" + end_time

    	tweet_fields = tweet_fields + end_field
    if start_time:
    	start_field = "&start_time=" + start_time

    	tweet_fields = tweet_fields + start_field

    if location != None:
    	country_fields = " -place:\"" + location + '\"'

    if next_token:
    	tweet_fields += "&next_token=" + next_token

    tweet_fields += "&max_results=" + str(numberOfTweets)
    #query += country_fields






    url = "https://api.twitter.com/2/tweets/search/recent?query={}&{}".format(
        query, tweet_fields
    )

    return url


def create_headers(bearer_token):
    headers = {"Authorization": "Bearer {}".format(bearer_token)}
    return headers






def connect_to_endpoint(url, headers):
    response = requests.request("GET", url, headers=headers)
    print(response.status_code)
    if response.status_code != 200:
        raise Exception(response.status_code, response.text)
    return response.json()


def searchTweets(numberOfTweets, searchTerm, location, start_time, end_time):
	bearer_token = bearerToken
	thelist = []
	next_token  = None
	while numberOfTweets > 0:
		curNum = min(100,numberOfTweets)
		url = create_url("BTS",location,start_time,end_time, curNum,next_token)
		headers = create_headers(bearer_token)
		response = connect_to_endpoint(url, headers)

		#response = json.dumps(json_response, indent=4, sort_keys=True)
		next_token = response["meta"]["next_token"]
		thelist +=  response["data"]
		numberOfTweets -= curNum
	return thelist

print(searchTweets(20,"BTS", None, None, "2021-03-24T09:55:06-00:00"))